### PR TITLE
Improve applications endpoint

### PR DESCRIPTION
**Pull request by billiford**
The`/applications/*` endpoints are probably the most difficult endpoints to work with and are prone to the most issues. This PR mainly fixes timeouts associated with listing resources for the Clusters page in Spinnaker. Previously, if a cluster did not exist, the following endpoints would timeout with these times:
```
/applications/*/serverGroupManagers: 30s
/applications/*/loadBalancers: 1m
/applications/*/serverGroups: 30s
```
This is *bad*. It means that if any cluster is not available for a given application then the Clusters page basically shows nothing for a while, until the caching on Gate catches up.

This is because listing server group managers would attempt to list all Deployments then ReplicaSets serially. When it failed on Deployments it would return (30 second timeout). Setting the List Timeout in the Kube Client List Options to 10 seconds has *never* worked. 

Load Balancers would time out at one minute, because it would try to list Services, fail after thirty seconds, *then continue* and try to list Ingresses and fail after thirty seconds.

Listing Server Groups failed in a similar manner to listing Server Group Managers, but failed when listing pods.

**In summary, this PR does the following**
- adds the ability to list resources with a context (where we set the timeout to be 10 seconds in this scenario)
- passes a copy of the gin context to go routines (see https://github.com/gin-gonic/gin#goroutines-inside-a-middleware)
- lists all resources for all of these endpoints concurrently
- uses the same function for all list operations to simplify the code
- adds additional comments to the `applications.go` file as this file is notoriously tough to work with

This should make it so that no matter how many clusters are not available for a given application, the Clusters page in Spinnaker will take no longer than ~10 seconds to load. Make sure to expand the `applications.go` file when reviewing.

### CODE REVIEW

**Comment by billiford** [comment]
There is a lot of repeated code here... let me try to fix that.

**Review by dmrogers7 — APPROVED** [review]
👍
